core/shared/src/main/scala/scalaz/zio/Fiber.scala

```diff
@@ -51,43 +61,25 @@ trait Fiber[+E, +A] { self =>
    */
   def interrupt0(ts: List[Throwable]): IO[Nothing, Unit]
 
-  /**
-   * Add an exit handler for when the fiber terminates and receive information
-   * on whether the fiber terminated normally, with unhandled error, or with
-   * exception.
-   *
-   * The specified action will be invoked after the fiber has finished running
-   * (including all finalizers). If the specified action throws an exception,
-   * it will be reported to the parent fiber's unhandled error handler.
-   */
-  def onComplete(f: ExitResult[E, A] => IO[Nothing, Unit]): IO[Nothing, Unit]
-
   /**
    * Zips this fiber with the specified fiber, combining their results using
    * the specified combiner function. Both joins and interruptions are performed
    * in sequential order from left to right.
    */
   final def zipWith[E1 >: E, B, C](that: => Fiber[E1, B])(f: (A, B) => C): Fiber[E1, C] =
     new Fiber[E1, C] {
-      def join: IO[E1, C] =
-        self.join.seqWith(that.join)(f)
+      def observe: IO[Nothing, ExitResult[E1, C]] =
+        self.observe.seqWith(that.observe) {
+          case (ExitResult.Completed(a), ExitResult.Completed(b))   => ExitResult.Completed(f(a, b))
+          case (ExitResult.Failed(e, ts), rb)                       => ExitResult.Failed(e, combine(ts, rb))
+          case (ExitResult.Terminated(ts), rb)                      => ExitResult.Terminated(combine(ts, rb))
+          case (ExitResult.Completed(_), ExitResult.Failed(e, ts))  => ExitResult.Failed(e, ts)
+          case (ExitResult.Completed(_), ExitResult.Terminated(ts)) => ExitResult.Terminated(ts)
```

Comment:
It seems there are missing cases here? I'm not sure why the compiler is not picking them up.

Reply:
After patterns of the previous two lines fail, we know fiber `a` is competed and fiber `b` is in an error state (the happy path being the first match).